src/sentry/db/models/fields/hybrid_cloud_foreign_key.py

```diff
@@ -0,0 +1,59 @@
+"""
+A 'foreign key' which is not enforced in the local database, but triggers eventually consistent delete work in the
+presence of RegionTombstone ControlTombstone model objects through the tasks/deletion.py script.
+
+This column is, in fact, an integer, and cannot enforce integrity directly.  It's main purpose to simply to track
+the need for cleanup in the case of a resource which is tracked in an opposing SiloMode.
+"""
+from __future__ import annotations
+
+from enum import IntEnum
+
+from django.db import models
+
+__all__ = "HybridCloudForeignKey"
+
+from typing import Any, Tuple
+
+from django.apps import apps
+
+
+class HybridCloudForeignKeyCascadeBehavior(IntEnum):
+    CASCADE = 1
+    SET_NULL = 2
+
+
+class HybridCloudForeignKey(models.BigIntegerField):  # type: ignore
```

Comment:
Maybe a bit pedantic, but any chance we could actually split this change out into a few PRs.

1. I think introducing this new field type probably could be a PR on it's own
2. Followed by the new deletions code that generically handles deletions of these relations
3. And finally the actual changes that use both of these new services features to update the SavedSearch → User FK split

I believe this would make reviewing of each of the parts a little easier. What do you think?

Reply:
It's definitely possible to do that, but sincerely worry that seeing any of these parts in isolation results in reviews that "pull on an empty chain" and result in yak shaving.


The HybridCloudForeignKey field is completely irrelevant in isolation -- it does absolutely nothing business logic wise.  Originally, hybrid cloud foreign key meta data was registered in a manager away from the model definitions, so that testing and registering this metadata could be separated from doing the split migration.

  The feedback I received was that having an explicit new field where that was registered was more clear, but having the metadata registered in two places also doesn't make sense, so I settled on having the model be the source of truth for deletion processing metadata.

 However, that decision and its implication is only contextualized by seeing the fact that if the model IS the source of truth for the deletion metadata, then it's ALSO responsible for producing migrations.

The fact that this field exists is only contextualized by all 3 parts (migration, metadata, deletion).  I worry that, despite the large context here, any further splitting is "easier" reviews that are also "less useful".

Reply:
Yeah I do agree, the context for the changes is important, I would probably recommend reviews to look at all 3 PRs to fully understand the usages.

Reply:
Yes, but that also... splits comments across 3 PRs, and we have to look at all 3 parts.  Compromise: does it help it I reorganize the history of this 1 PR into 3 discrete commits?

Reply:
It would help yes!

Reply:
> splits comments across 3 PRs, and we have to look at all 3 parts

That's fine right, the comments should be for each portion of the code.

Also

> The HybridCloudForeignKey field is completely irrelevant in isolation -- it does absolutely nothing business logic wise. Originally, hybrid cloud foreign key meta data was registered in a manager away from the model definitions, so that testing and registering this metadata could be separated from doing the split migration.
> 
> The feedback I received was that having an explicit new field where that was registered was more clear, but having the metadata registered in two places also doesn't make sense, so I settled on having the model be the source of truth for deletion processing metadata.
> 
> However, that decision and its implication is only contextualized by seeing the fact that if the model IS the source of truth for the deletion metadata, then it's ALSO responsible for producing migrations.

It sounds like the FK should have more documentation to make clear it's intended usage. Without that users of the new field type have to rely on looking at existing usage.